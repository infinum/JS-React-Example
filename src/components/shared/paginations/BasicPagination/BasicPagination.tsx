import { FC } from 'react';
import { Button, HStack } from '@chakra-ui/react';

export interface IBasicPagination {
	current: number;
	total: number;
	onPrev?: () => void;
	hasPrev?: boolean;
	onNext?: () => void;
	hasNext?: boolean;
}

export const BasicPagination: FC<IBasicPagination> = ({ hasPrev, hasNext, onPrev, onNext, current, total }) => {
	return (
		<HStack p={50} spacing={2} justify="center">
			<Button isDisabled={!hasPrev} onClick={onPrev}>
				Previous
			</Button>
			<Button as="div">{`${current}/${total}`}</Button>
			<Button isDisabled={!hasNext} onClick={onNext}>
				Next
			</Button>
		</HStack>
	);
};
