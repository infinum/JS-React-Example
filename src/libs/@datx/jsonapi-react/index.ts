export { DatxProvider } from './context';
export { useDatx } from './hooks/use-datx';
export { useResource } from './hooks/use-resource';
export { useResourceList } from './hooks/use-resource-list';
export { Resource } from './Resource';
export { Client } from './Client';
export { deapify, apify, compare } from './utils';
export * from './types';
