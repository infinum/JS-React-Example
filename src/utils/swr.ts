import { Model } from 'datx';

/*
  Inspired by https://github.com/lukeed/dequal/blob/master/src/lite.js
*/
export const swrComparator = (a: unknown, b: unknown) => {
  if (a === b) {
    return true;
  }

  if (a && b && a.constructor === b.constructor) {
    let length: number;

    if (Array.isArray(a) && Array.isArray(b)) {
      // eslint-disable-next-line no-cond-assign
      if ((length = a.length) === b.length) {
        while (length-- && swrComparator(a[length], b[length]));
      }
      return length === -1;
    }

    if (a instanceof Model && b instanceof Model) {
      return swrComparator(a.meta.snapshot, b.meta.snapshot);
    }

    if (!a.constructor || typeof a === 'object') {
      length = 0;
      // eslint-disable-next-line guard-for-in, no-restricted-syntax
      for (const key in a as Record<string, unknown>) {
        if (
          Object.prototype.hasOwnProperty.call(a, key) &&
          ++length &&
          !Object.prototype.hasOwnProperty.call(b, key)
        ) {
          return false;
        }
        if (!(key in (b as Record<string, unknown>)) || !swrComparator(a[key], b[key])) {
          return false;
        }
      }

      return Object.keys(b).length === length;
    }
  }

  // eslint-disable-next-line no-self-compare
  return a !== a && b !== b;
};
