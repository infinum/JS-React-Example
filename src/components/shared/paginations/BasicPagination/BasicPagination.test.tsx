import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import mockRouter from 'next-router-mock';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import { render, screen } from 'test-utils';
import { BasicPagination, getPagination } from './BasicPagination';
import { CollectionResponse } from '@datx/swr';

describe('BasicPagination', () => {
	const defaultPagination = {
		hasNext: false,
		hasPrevious: false,
		currentPage: 1,
		totalPages: 1,
	};

	it('should matches snapshot with falsy values', () => {
		const { asFragment } = render(<BasicPagination pagination={defaultPagination} />);

		expect(asFragment()).toMatchSnapshot('falsy values');
	});

	it('should matches snapshot with truthy values', () => {
		const { asFragment } = render(
			<BasicPagination
				pagination={{
					hasNext: true,
					hasPrevious: true,
					currentPage: 2,
					totalPages: 4,
				}}
			/>
		);

		expect(asFragment()).toMatchSnapshot('truthy values');
	});

	it('should be accessible', async () => {
		const { container } = render(<BasicPagination pagination={defaultPagination} />);

		const results = await axe(container);

		expect(results).toHaveNoViolations();
	});

	it('should set page query param on action', async () => {
		await mockRouter.push('/');
		const user = userEvent.setup();

		const { rerender } = render(
			<BasicPagination pagination={{ ...defaultPagination, hasNext: true, totalPages: 4 }} />,
			{ wrapper: MemoryRouterProvider }
		);

		const nextLink = screen.queryByRole('link', { name: /next/i });

		expect(nextLink).toBeInTheDocument();
		await user.click(nextLink as HTMLElement);

		expect(mockRouter).toMatchObject(
			expect.objectContaining({
				asPath: '/?page=2',
				pathname: '/',
				query: { page: '2' },
			})
		);

		rerender(
			<BasicPagination pagination={{ ...defaultPagination, hasPrevious: true, currentPage: 2, totalPages: 4 }} />
		);

		const prevLink = screen.queryByRole('link', { name: /prev/i });

		expect(prevLink).toBeInTheDocument();
		await user.click(prevLink as HTMLElement);

		expect(mockRouter).toMatchObject(
			expect.objectContaining({
				asPath: '/?page=1',
				pathname: '/',
				query: { page: '1' },
			})
		);
	});

	it('should keep other query params on action', async () => {
		await mockRouter.push('/?query=param');
		const user = userEvent.setup();

		render(<BasicPagination pagination={{ ...defaultPagination, hasNext: true, totalPages: 4 }} />, {
			wrapper: MemoryRouterProvider,
		});

		const nextLink = screen.queryByRole('link', { name: /next/i });

		expect(nextLink).toBeInTheDocument();

		await user.click(nextLink as HTMLElement);

		expect(mockRouter).toMatchObject(
			expect.objectContaining({
				asPath: '/?query=param&page=2',
				pathname: '/',
				query: { page: '2', query: 'param' },
			})
		);
	});
});

describe('getPagination', () => {
	it('should return empty object with no response', () => {
		expect(getPagination()).toEqual({});
	});

	it('should return pagination object from response', () => {
		expect(
			getPagination(
				new CollectionResponse({
					data: {
						meta: {
							currentPage: 1,
							totalPages: 1,
						},
					},
					status: 200,
				})
			)
		).toEqual({
			currentPage: 1,
			totalPages: 1,
			hasNext: false,
			hasPrevious: false,
		});
	});
});
