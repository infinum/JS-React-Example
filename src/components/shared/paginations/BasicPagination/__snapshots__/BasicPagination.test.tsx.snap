// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicPagination should matches snapshot with falsy values: falsy values 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 50px;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 0px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: var(--chakra-space-2);
  margin-inline-start: var(--chakra-space-2);
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
  line-height: 1.2;
  border-radius: var(--chakra-radii-md);
  font-weight: var(--chakra-fontWeights-semibold);
  transition-property: var(--chakra-transition-property-common);
  transition-duration: var(--chakra-transition-duration-normal);
  height: var(--chakra-sizes-10);
  min-width: var(--chakra-sizes-10);
  font-size: var(--chakra-fontSizes-md);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  background: var(--chakra-colors-gray-100);
}

.emotion-1:focus-visible,
.emotion-1[data-focus-visible] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-1:disabled,
.emotion-1[disabled],
.emotion-1[aria-disabled=true],
.emotion-1[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: var(--chakra-shadows-none);
}

.emotion-1:hover,
.emotion-1[data-hover] {
  background: var(--chakra-colors-gray-200);
}

.emotion-1:hover:disabled,
.emotion-1[data-hover]:disabled,
.emotion-1:hover[disabled],
.emotion-1[data-hover][disabled],
.emotion-1:hover[aria-disabled=true],
.emotion-1[data-hover][aria-disabled=true],
.emotion-1:hover[data-disabled],
.emotion-1[data-hover][data-disabled] {
  background: var(--chakra-colors-gray-100);
}

.emotion-1:active,
.emotion-1[data-active] {
  background: var(--chakra-colors-gray-300);
}

<div
    class="chakra-stack emotion-0"
  >
    <button
      class="chakra-button emotion-1"
      disabled=""
      href="[object Object]"
      type="button"
    >
      Previous
    </button>
    <div
      class="chakra-button emotion-1"
    >
      1/1
    </div>
    <button
      class="chakra-button emotion-1"
      disabled=""
      href="[object Object]"
      type="button"
    >
      Next
    </button>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`BasicPagination should matches snapshot with truthy values: truthy values 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 50px;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 0px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: var(--chakra-space-2);
  margin-inline-start: var(--chakra-space-2);
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
  line-height: 1.2;
  border-radius: var(--chakra-radii-md);
  font-weight: var(--chakra-fontWeights-semibold);
  transition-property: var(--chakra-transition-property-common);
  transition-duration: var(--chakra-transition-duration-normal);
  height: var(--chakra-sizes-10);
  min-width: var(--chakra-sizes-10);
  font-size: var(--chakra-fontSizes-md);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  background: var(--chakra-colors-gray-100);
}

.emotion-1:focus-visible,
.emotion-1[data-focus-visible] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-1:disabled,
.emotion-1[disabled],
.emotion-1[aria-disabled=true],
.emotion-1[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: var(--chakra-shadows-none);
}

.emotion-1:hover,
.emotion-1[data-hover] {
  background: var(--chakra-colors-gray-200);
}

.emotion-1:hover:disabled,
.emotion-1[data-hover]:disabled,
.emotion-1:hover[disabled],
.emotion-1[data-hover][disabled],
.emotion-1:hover[aria-disabled=true],
.emotion-1[data-hover][aria-disabled=true],
.emotion-1:hover[data-disabled],
.emotion-1[data-hover][data-disabled] {
  background: var(--chakra-colors-gray-100);
}

.emotion-1:active,
.emotion-1[data-active] {
  background: var(--chakra-colors-gray-300);
}

<div
    class="chakra-stack emotion-0"
  >
    <a
      class="chakra-button emotion-1"
      href="?page=1"
    >
      Previous
    </a>
    <div
      class="chakra-button emotion-1"
    >
      2/4
    </div>
    <a
      class="chakra-button emotion-1"
      href="?page=3"
    >
      Next
    </a>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
